### PR TITLE
Feature/1708 affiliate link owner info show

### LINKED ISSUE #1708: [Affiliate] Show a user info about affiliate link owner
Currently, we do not indicate which affiliate link user is following and under which link the 1st trade was done. 

Even if we follow the same link once again, we will see the common message 'Invalid affiliate code: The currently connected wallet has traded before or is already part of the affiliate program.'

So it would be nice to add an info about affiliate link user followed into the Profile details page. 
Also, show a different banner to a user when he/she follow the same link after a successful trade. 
Something like 'Affiliate program: You have already traded with this affiliate code' or so

### PR DESCRIPTION
**Pull request by maria-vslvn**
# Summary

<<if there's an issue>>Fixes #1708 

add and modify the affiliate link owner info display ate the profile page

  # To Test

1. Open the page `Profile`
Checkout the new display and updates

### CODE REVIEW

**Review by elena-zh — CHANGES_REQUESTED** [review]
Hey @maria-vslvn , 

some issues: 
1. It would be nice to hide the field when a user is not connected to the Mainnet, as referral data is available only there
mainnet

2. It would be great to hide the field if a user has not traded before
do not show when no trades

3. No need to show this link to a user with trades, and these trades were done without a referral link

4. Referral link can't be connected address. The value is the address that is uploaded to ipfs when a user follows a referral link

5. Seems that the fields are not placed in 1 horizontal line on the form
not in line

6. Mobile and tablet views are changed in this PR. It would be great to leave them as they are currently deployed in Prod
mobile
tablet

**Comment by ramirotw** [comment]
@maria-vslvn you should pick the referral address by using the hook `useReferralAddress`.

And btw there's a typo in `Reffered` -> `Referred`

**Review by ramirotw — APPROVED** [review]
👍

File: src/custom/pages/Profile/index.tsx [hunk 1]
```diff
@@ -28,9 +28,12 @@ import NotificationBanner from 'components/NotificationBanner'
 import { SupportedChainId as ChainId } from 'constants/chains'
 import AffiliateStatusCheck from 'components/AffiliateStatusCheck'
 import { useHasOrders } from 'api/gnosisProtocol/hooks'
+import { shortenAddress } from '@src/utils'
```
Comment (alfetopito) [inline src/custom/pages/Profile/index.tsx:31]: Don't use `@src` imports.

Probably will work without it (I have not tested it)

```suggestion
import { shortenAddress } from 'utils'
```

**Review by alfetopito — APPROVED** [review]
Minor comment, otherwise looking good to me.

Would just make sure Elena's comments are addressed

**Comment by elena-zh** [comment]
Hi @maria-vslvn , cases 1-4 from the comment above https://github.com/gnosis/cowswap/pull/2026#pullrequestreview-837335359 are still not fixed. 
The purpose of the current issue is to show a user an address he/she has followed, and under which he/she has done a trade. This value  (referral address) should be taken from the BE.
In this PR I see that you are trying to show the new referral link, that is currently active, and is stored in the local storage. 
I assume, that is why all these issues are still not fixed. 

Thanks.

**Review by elena-zh — CHANGES_REQUESTED** [review]
We still do not indicate under which link the 1st trade was done.

**Comment by matextrem** [comment]
@ramirotw  @alfetopito  Which is the endpoint needed to retrieve the referrer user? `Trades` enpoint does not return any information related to that.

**Comment by ramirotw** [comment]
> @ramirotw @alfetopito Which is the endpoint needed to retrieve the referrer user? `Trades` enpoint does not return any information related to that.

I believe the way we could do that is:
1. get the user first trade by using the /trades endpoint
2. using the orderId from that trade query the /orders/{uid}
3. using the appData from the order response, we could use the decode helper @alfetopito implemented in https://codepen.io/pacara/pen/BadZgjg (this should be part of the sdk btw)
4. extract the referrer address from there

does that makes sense @alfetopito or we have a simpler and better way to do it?

**Comment by alfetopito** [comment]
> > @ramirotw @alfetopito Which is the endpoint needed to retrieve the referrer user? `Trades` enpoint does not return any information related to that.
> 
> I believe the way we could do that is:
> 
>     1. get the user first trade by using the /trades endpoint
> 
>     2. using the orderId from that trade query the /orders/{uid}
> 
>     3. using the appData from the order response, we could use the decode helper @alfetopito implemented in https://codepen.io/pacara/pen/BadZgjg (this should be part of the sdk btw)
> 
>     4. extract the referrer address from there
> 
> 
> does that makes sense @alfetopito or we have a simpler and better way to do it?

Yeah, that sounds about right.
Don't think there's an easier way right now.

**Comment by matextrem** [comment]
I have read and agree to the CLA

**Comment by elena-zh** [comment]
Hey @matextrem , not I can see a duplicating 'View all orders' link in any network when a user has trades. 
However, I can't see a referee data on the profile page
image

**Comment by matextrem** [comment]
@elena-zh Removed one view all orders link

Regarding the other point, I can see it. Take into account that it will only shown once the referred address is fetched.

image

**Comment by elena-zh** [comment]
Hey @matextrem , great!

The only thing I should notice that it would be nice to aggregate affiliate data for the referee for Barn and Prod environments and show it. I mean, this account on the image below has a referral link that was shared in Prod server. However, it is not displayed in the barn/dev environment 
image
Account's trades: https://protocol-explorer.staging.gnosisdev.com/address/0xdEB73112e933b7e7E874BaBE33D2EbECc49f0514
Referral address: 0x6adE2eA11D5a119D8b431ef03DcD5401B733CEbe

Another my question is related to the mobile view: should we show this link above the section's title or we can move it down. @alongoni , WDYT?

**Comment by matextrem** [comment]
Regarding the first point from Elena comment and talking to Rami, a solution for this would be trying to getTrades pointing to the default env (i.e: barn). If getTrades does not returns anything, try with prod environment.

WDTY is the best approach to handle it @alfetopito ?

CC: @ramirotw

File: src/custom/pages/Profile/index.tsx [hunk 2]
```diff
@@ -75,7 +79,17 @@ export default function Profile() {
       <Wrapper>
         <GridWrap>
           <CardHead>
-            <Title>Affiliate Program</Title>
+            {account && refAddress.value && (
+              <FlexRow>
```
Comment (alongoni) [inline src/custom/pages/Profile/index.tsx:83]: It should be like this (without adding extra space upper the Affiliate Program title)
image

Comment (alongoni) [reply src/custom/pages/Profile/index.tsx:83]: In mobile, I agree with Elena, we should list "Referred by: ... " below the "Las updated..." data.

**Review by alongoni — CHANGES_REQUESTED** [review]

**Review by alongoni — APPROVED** [review]
Looks great!
image
image

**Comment by alfetopito** [comment]
> Regarding the first point from Elena comment and talking to Rami, a solution for this would be trying to getTrades pointing to the default env (i.e: barn). If getTrades does not returns anything, try with prod environment.
> 
> WDTY is the best approach to handle it @alfetopito ?
> 
> CC: @ramirotw

I was sure something like that had already been implemented for the affiliate program, before we upload it to IPFS, but looking at the code right now I couldn't find it.

Did we just talk about it but did not actually implement it?

Anyway, I think this would make sense for both but could be a follow up PR

**Comment by matextrem** [comment]
@alfetopito You mean this ticket? https://github.com/gnosis/cowswap/issues/1819. @elena-zh shared it to me but I was unsure if it was the same case.

**Comment by alfetopito** [comment]
@matextrem I've raised this internally with the backend team to check for an alternative solution.

Will let you know when I get an answer

**Comment by matextrem** [comment]
@elena-zh @alfetopito updated!

**Comment by matextrem** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by matextrem** [comment]
recheckcla

**Comment by elena-zh** [comment]
Hey @matextrem , seems that the last your commit was not properly merged as I still can't see a referree for an account with trades in the Prod server
image
image

@alongoni , WDYT about left-aligning all these elements in a mobile view like we do for the tablet view?
image
image

Also, it would be nice to make have a common style:  currently, icons, fields' labels have different sizes, colors, etc.

File: src/custom/api/gnosisProtocol/api.ts [hunk 3]
```diff
@@ -425,6 +484,45 @@ export async function getTrades(params: GetTradesParams): Promise<TradeMetaData[
...
 
+export async function getAllTrades(params: GetTradesParams): Promise<TradeMetaData[]> {
+  const { chainId, owner, limit, offset } = params
+  const qsParams = stringify({ owner, limit, offset })
+  const stagingUrl =
+    process.env[`REACT_APP_API_URL_STAGING_${ChainId[chainId]}`] ||
+    `https://barn.api.cow.fi/${ChainId[chainId].toLowerCase()}/api`
+  const prodUrl =
+    process.env[`REACT_APP_API_URL_PROD_${ChainId[chainId]}`] ||
+    `https://api.cow.fi/${ChainId[chainId].toLowerCase()}/api`
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/api/gnosisProtocol/api.ts:495]: Probably a const for each: 

```ts
const BARN_API_ENDPOINTS_PER_NETWORK: Record<ChainId, string> = {}
const PROD_API_ENDPOINTS_PER_NETWORK: Record<ChainId, string> = {}
```

**Comment by alongoni** [comment]
Regarding styles:
image
@matextrem you can check the CSS here:
https://www.figma.com/file/bcoZf5RIgVKDItdOaTZq18/?node-id=1759%3A2

Not sure yet about left-aligning all elements (Referred by + Last updated + View all orders). I'll check it in a prototype on figma.

File: src/custom/api/gnosisProtocol/api.ts [hunk 4]
```diff
@@ -381,6 +404,42 @@ export async function getOrder(chainId: ChainId, orderId: string): Promise<Order
...
+      await _getAllOrder(chainId, `/orders/${orderId}`, stagingUrl),
+      await _getAllOrder(chainId, `/orders/${orderId}`, prodUrl),
+    ])
+
+    const response = results
+      .filter(({ status }) => status === 'fulfilled')
+      .filter((res) => (res as PromiseFulfilledResult<any>).value.ok)
+      .map((p) => (p as PromiseFulfilledResult<any>).value)
+
+    const [stgResponse, prodResponse] = response
```
[the remark this answers is not in the document]
Comment (matextrem) [reply src/custom/api/gnosisProtocol/api.ts:398]: Worst case scenario, `stgResponse` and `prodResponse` would be `undefined`. They are validated later in the code so there wouldn't be any problem I think.

The thing is `Promise.allSettled` returns `PromiseSettledResult` which is not exported so you have to first filter by `fulfilled` in order to access to `value` key https://stackoverflow.com/a/67533949

File: src/custom/api/gnosisProtocol/api.ts [hunk 5]
```diff
@@ -425,6 +484,45 @@ export async function getTrades(params: GetTradesParams): Promise<TradeMetaData[
...
+
+    if (!response.length) {
+      const [stgErrResponse, prodErrResponse] = results as PromiseFulfilledResult<any>[]
```
[the remark this answers is not in the document]
Comment (matextrem) [reply src/custom/api/gnosisProtocol/api.ts:477]: There seems to be a problem with `Promise.allSettled `and typing (https://stackoverflow.com/a/67533949) . I tried the solution suggested there but I was not able to make it work with our current `eslint` configuration so I have to type those cases with `any`

**Comment by elena-zh** [comment]
Hey @matextrem , great changes!

The last thing that I have noticed today is that the account referee info  is not auto-removed when switch to another account without trades/referral code. See the video: https://watch.screencastify.com/v/xHXmzYwUEKS80xTjMmWU

Could you please fix this?

**Comment by matextrem** [comment]
@elena-zh Updated!

**Review by elena-zh — APPROVED** [review]
LGTM now!

**Comment by alfetopito** [comment]
As was discussed in the last sync meeting, we decided to not proceed with this feature.

Thanks for the effort everyone, but unfortunately it won't go live.

**Pull request closed without merging**